src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/nodes/ExecuteTopLevelContextNode.java

```diff
@@ -133,82 +140,76 @@ private static ContextObject returnTo(final ContextObject activeContext, final A
-                }
-            }
             final AbstractSqueakObject currentSender = context.getSender();
             if (currentSender instanceof final ContextObject o) {
                 context = o;
+            } else {
+                return sendCannotReturn(activeContext, returnValue);
             }
         }
+        // Evaluate unwind-marked blocks on sender chain.
         context = senderContext;
         while (context != targetContext) {
-            final AbstractSqueakObject currentSender = context.getSender();
-            if (currentSender instanceof final ContextObject o) {
-                context.terminate();
-                context = o;
-            } else { // TODO: this might need to be handled by a cannotReturn send.
-                LogUtils.SCHEDULING.warning("Unwind error: sender of " + context + " is nil, unwinding towards " + targetContext + " with return value: " + returnValue);
-                break;
-            }
-        }
-        targetContext.push(returnValue);
-        return targetContext;
-    }
-
-    @TruffleBoundary
-    private ContextObject commonReturn(final ContextObject startContext, final ContextObject targetContext, final Object returnValue) {
-        /* "make sure we can return to the given context" */
-        if (!targetContext.hasClosure() && !targetContext.canBeReturnedTo()) {
-            if (startContext == targetContext) {
-                throw returnToTopLevel(targetContext, returnValue);
-            }
-            return sendCannotReturn(startContext, returnValue);
-        }
-        /*
-         * "If this return is not to our immediate predecessor (i.e. from a method to its sender, or
-         * from a block to its caller), scan the stack for the first unwind marked context and
-         * inform this context and let it deal with it. This provides a chance for ensure unwinding
-         * to occur."
-         */
-        AbstractSqueakObject contextOrNil = startContext;
-        while (contextOrNil != targetContext) {
-            if (!(contextOrNil instanceof final ContextObject context)) {
-                /* "error: sender's instruction pointer or context is nil; cannot return" */
-                assert contextOrNil == NilObject.SINGLETON;
-                return sendCannotReturn(startContext, returnValue);
-            }
-            assert !context.isPrimitiveContext();
             if (context.getCodeObject().isUnwindMarked()) {
-                assert !context.hasClosure();
-                /* "context is marked; break out" */
-                return sendAboutToReturn(startContext, returnValue, context);
+                try {
+                    // TODO: make this better. Clearing the modified sender permits virtualization
+                    // of aboutToReturn
+                    context.clearModifiedSender();
+                    AboutToReturnNode.create(context.getCodeObject()).executeAboutToReturn(context.getTruffleFrame(), nlr);
```

Comment:
`commonNLReturn()` is behind a `@TruffleBoundary` so excluded from compilation where values may be virtualized. That's also why we previously used `sendAboutToReturnNode` here. Were you trying to optimize this here or can we use `sendAboutToReturn()` and the node again?

Reply:
What I was trying to do here was to force the use of AboutToReturnNode>>doAboutToReturnVirtualized because I was unsure whether we could mix the virtualized and message-based implementations of aboutToReturn.

It seems to me that once we start using doAboutToReturnVirtualized, the homeContext no longer has a sender and so the message would fail.